src/librustc_data_structures/profiling.rs

```diff
@@ -585,3 +601,82 @@ fn get_resident() -> Option<usize> {
         }
     }
 }
+
+fn pass_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| profiler.alloc_string(llvm_name))
+}
+
+fn ir_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| {
+        let demangled_ir_name = rustc_demangle::demangle(llvm_name).to_string();
+        profiler.alloc_string(demangled_ir_name.as_str())
+    })
+}
+
+fn llvm_args_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    pass_name: &str,
+    ir_name: &str,
+) -> EventId {
+    let pass_name = pass_name_to_string_id(profiler, string_cache, pass_name);
+    let mut components = vec![StringComponent::Ref(pass_name)];
+    // handle that LazyCallGraph::SCC is a comma separated list within parentheses
+    let parentheses: &[_] = &['(', ')'];
+    let trimed = ir_name.trim_matches(parentheses);
+    for part in trimed.split(',') {
+        let ir_name = ir_name_to_string_id(profiler, string_cache, part);
+        components.push(StringComponent::Value(SEPARATOR_BYTE));
+        components.push(StringComponent::Ref(ir_name));
+    }
+    EventId::from_label(profiler.alloc_string(components.as_slice()))
+}
+
+pub struct LlvmSelfProfiler<'a> {
```

Comment:
I think `LlvmSelfProfiler` and all of the related code should be moved to `librustc_codegen_llvm`, e.g. into a file `librustc_codegen_llvm/back/profiling.rs`.

Reply:
will look in to doing this next week

Reply:
No rush

Reply:
have started looking at this and have a question.
now I have a get_llvm_selfprofiler() function in SelfProfilerRef but that do not work after the move. but some how I need to get a selfprofiler to save in LlvmSelfProfiler and I need to check if the llvm event is wanted. shall I make profiler public(or a get function) and add a llvm_recording_enabled in SelfProfilerRef? or shall I have a get_profiler_when_llvm_recording_enabled that returns a option? I feel like neither is a good solution.